### Funciones vectorizadas y limpieza de datos

#### Lectura de CSV's

La sesión pasada aprendimos a leer archivos en formato JSON. El formato JSON, que es muy parecido a los diccionarios de Python, es sólo uno de los tantos formatos con los que nos vamos a topar.

Los CSVs pertenecen a una clase de formatos donde las columnas de nuestra tabla se delimitan usando lo que se llama un separador. CSV significa Comma-Separated Values y como bien imaginarás significa que se una una coma (,) para separar las columnas. Un CSV se ve así:

Id,Date,AveragePrice,Total Volume,4046,4225,4770,Total Bags,Small Bags,Large Bags,XLarge Bags,type,year,region
0,2015-12-27,1.33,64236.62,1036.74,54454.85,48.16,8696.87,8603.62,93.25,0.0,conventional,2015,Albany
1,2015-12-20,1.35,54876.98,674.28,44638.81,58.33,9505.56,9408.07,97.49,0.0,conventional,2015,Albany
2,2015-12-13,0.93,118220.22,794.7,109149.67,130.5,8145.35,8042.21,103.14,0.0,conventional,2015,Albany
3,2015-12-06,1.08,78992.15,1132.0,71976.41,72.58,5811.16,5677.4,133.76,0.0,conventional,2015,Albany
4,2015-11-29,1.28,51039.6,941.48,43838.39,75.78,6183.95,5986.26,197.69,0.0,conventional,2015,Albany
5,2015-11-22,1.26,55979.78,1184.27,48067.99,43.61,6683.91,6556.47,127.44,0.0,conventional,2015,Albany
6,2015-11-15,0.99,83453.76,1368.92,73672.72,93.26,8318.86,8196.81,122.05,0.0,conventional,2015,Albany

En un archivo de texto donde cada fila de nuestra tabla tiene su propia línea y donde los valores de cada columna se delimitan usando una coma (,). Leer archivos .csv usando pandas es muy fácil. Lo único que tienes que hacer es lo siguiente:

In [1]:
import pandas as pd

In [71]:
df = pd.read_csv('../Datasets/melbourne_housing-raw.csv', sep = ',')
df

Suburb               Address  Rooms Type      Price Method  \
0      Abbotsford         68 Studley St      2    h        NaN     SS   
1      Abbotsford          85 Turner St      2    h  1480000.0      S   
2      Abbotsford       25 Bloomburg St      2    h  1035000.0      S   
3      Abbotsford    18/659 Victoria St      3    u        NaN     VB   
4      Abbotsford          5 Charles St      3    h  1465000.0     SP   
...           ...                   ...    ...  ...        ...    ...   
19735     Windsor       201/152 Peel St      2    u   560000.0     PI   
19736     Wollert       60 Saltlake Bvd      3    h   525300.0      S   
19737  Yarraville           2 Adeney St      2    h   750000.0     SP   
19738  Yarraville       54 Pentland Pde      6    h  2450000.0     VB   
19739  Yarraville  10/127 Somerville Rd      3    t   645000.0     SP   

             SellerG        Date  Distance  Postcode  ...  Bathroom  Car  \
0             Jellis   3/09/2016       2.5    3067.0  ...       1.0  1.0   
1             Biggin   3/12/2016       2.5    3067.0  ...       1.0  1.0   
2             Biggin   4/02/2016       2.5    3067.0  ...       1.0  0.0   
3             Rounds   4/02/2016       2.5    3067.0  ...       2.0  1.0   
4             Biggin   4/03/2017       2.5    3067.0  ...       2.0  0.0   
...              ...         ...       ...       ...  ...       ...  ...   
19735  hockingstuart  29/07/2017       4.6    3181.0  ...       1.0  1.0   
19736      Stockdale  29/07/2017      25.5    3750.0  ...       2.0  2.0   
19737  hockingstuart  29/07/2017       6.3    3013.0  ...       2.0  2.0   
19738        Village  29/07/2017       6.3    3013.0  ...       2.0  1.0   
19739            Jas  29/07/2017       6.3    3013.0  ...       1.0  1.0   

       Landsize  BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
0         126.0           NaN        NaN        Yarra -37.80140   144.99580   
1         202.0           NaN        NaN        Yarra -37.79960   144.99840   
2         156.0          79.0    1900.00        Yarra -37.80790   144.99340   
3           0.0           NaN        NaN        Yarra -37.81140   145.01160   
4         134.0         150.0    1900.00        Yarra -37.80930   144.99440   
...         ...           ...        ...          ...       ...         ...   
19735       NaN         585.0        NaN   Whittlesea -37.67681   145.00323   
19736       NaN         333.0        NaN      Darebin -37.75884   145.00264   
19737    1999.0         199.0     140.00      Darebin -37.75948   144.99615   
19738    2011.0         238.0     118.00         Hume -37.70322   144.88236   
19739    1980.0           0.0      66.32         Hume -37.69815   144.88019   

                  Regionname Propertycount  
0      Northern Metropolitan        4019.0  
1      Northern Metropolitan        4019.0  
2      Northern Metropolitan        4019.0  
3      Northern Metropolitan        4019.0  
4      Northern Metropolitan        4019.0  
...                      ...           ...  
19735  Southern Metropolitan        4380.0  
19736  Northern Metropolitan        2940.0  
19737   Western Metropolitan        6543.0  
19738   Western Metropolitan        6543.0  
19739   Western Metropolitan        6543.0  

[19740 rows x 21 columns]

pandas tiene un muy conveniente método llamado read_csv que nos permite leer archivos .csv directamente. Ese mismo método también puede ayudarnos a leer otros formatos con columnas delimitadas por otros separadores. Por ejemplo, podemos leer **.tsv ('tab-separated values')**, que son archivos donde cada columna está delimitada por un tab (indentación). Sólo basta con llamar el método con el argumento **sep=\t**.

¡Como ves, leer archivos tipo .csv es muy fácil!

#### Analisis exploratorio de datos

Muy bien, empecemos nuestra exploración con las técnicas que ya conocemos primero:

In [72]:
df.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/09/2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  4/02/2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.8014    144.9958  Northern Metropolitan   
1        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
2     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
3        NaN        Yarra  -37.8114    145.0116  Northern Metropolitan   
4     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [73]:
df.shape

(19740, 21)

In [74]:
df.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea', 'Lattitude',
       'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

In [75]:
df.dtypes

Suburb            object
Address           object
Rooms              int64
Type              object
Price            float64
Method            object
SellerG           object
Date              object
Distance         float64
Postcode         float64
Bedroom2         float64
Bathroom         float64
Car              float64
Landsize         float64
BuildingArea     float64
YearBuilt        float64
CouncilArea       object
Lattitude        float64
Longtitude       float64
Regionname        object
Propertycount    float64
dtype: object

Ok, aquí es posible determinar algunas cosas importantes:

- El dataset tiene 19740 filas y 21 columnas
- Podemos observar que la información que tiene el dataset es básicamente cierta descripción de las propiedades, datos de su ubicación y tipo y fecha de venta.
- Hay varias columnas float64, rooms es int64, y el resto son object.
- Tenemos algunos Nans en el dataset.

¿NaNs? ¿Qué quiero decir con eso? Bueno, los NaNs son valores Not a Number que básicamente son valores nulos en nuestro dataset. Estos valores pueden causar muchos problemas, ya que son valores nulos en columnas numéricas pero no podemos realizar operaciones matemáticas con ellos (son valores Not a Number, así que las matemáticas están fuera de las posibilidades). En esta sesión vamos a aprender a lidiar con estos valores, pero antes tenemos que aprender algo llamado funciones vectorizadas, que nos ayudará mucho durante esta exploración.

##### Aritmética con Series y Funciones vectorizadas

¿Recuerdas nuestras funciones map y filter? A esas funciones les pasábamos nosotros una función y una lista y nos regresaban una lista con los resultados de aplicarle la función a cada uno de los elementos en orden.

Las funciones vectorizadas funcionan muy parecido, pero están optimizadas para funcionar con arreglos de pandas y de numpy (si quieres saber más sobre numpy [lee esto](https://numpy.org/)). Si tomas un arreglo de pandas (es decir, una Serie) y le aplicas una función vectorizada la función se aplica a todo el arreglo elemento por elemento y te regresa un arreglo del mismo tamaño con el resultado de la aplicación.

Aplicar funciones de manera vectorizada a una Serie de pandas es facilísimo, incluso más fácil que usar la función map. Primero veamos que podemos realizar una operación aritmética con una Serie y la operación se aplicará automáticamente a todo el arreglo "elemento por elemento":

In [9]:
serie = pd.Series([3, 8, 4, 6, 2])
serie * 5

0    15
1    40
2    20
3    30
4    10
dtype: int64

Podemos realizar cualquier operación matemática y la aplicación se hará de la misma manera:

In [10]:
serie * 10

0    30
1    80
2    40
3    60
4    20
dtype: int64

In [11]:
(serie + 100) / 2

0    51.5
1    54.0
2    52.0
3    53.0
4    51.0
dtype: float64

¡Qué genial! No tenemos que usar map, ni que declarar una función, ni que usar lambda. Basta con una simple operación matemática. Pandas además está optimizado para funcionar de esta manera, así que la velocidad de aplicación es mucho mayor que la combinación de map con listas.

Otra manera de aplicar estas transformaciones es usando funciones vectorizadas. Por ejemplo, esto lo podemos hacer usando algunas funciones de numpy. Numpy es otra librería que es muy común entre los científicos de datos. Ofrece muchas herramientas para realizar cálculos numéricos a altas velocidades. No usaremos mucho esta librería en este módulo, pero es importante entender que se pueden utilizar funciones de numpy para aplicar funciones de manera vectorizada a Series de pandas. Para importar numpy hacemos lo siguiente:

In [12]:
import numpy as np

In [13]:
#La serie elevada al cuadrado con numpy
np.power(serie, 2)

0     9
1    64
2    16
3    36
4     4
dtype: int64

In [15]:
#Raíz cuadrada de la serie con numpy
np.sqrt(serie)

0    1.732051
1    2.828427
2    2.000000
3    2.449490
4    1.414214
dtype: float64

##### Agregaciones

Hay una variación de estas funciones vectorizadas llamadas agregaciones (o reducciones) que lo que hacen es tomar un arreglo, atravesarlo "elemento por elemento" y regresar un solo número que es un "resumen" del arreglo. Este "resumen" es justamente la agregación o reducción. Podemos aplicar estas funciones usando numpy o directamente desde una Serie o DataFrame de pandas. Para efectos prácticos, vamos a utilizar los métodos que vienen integrados directamente en pandas. Por ejemplo, podemos sumar todos los valores de una Serie de esta manera:

In [16]:
#Suma de todos los elementos de la serie
serie = pd.Series([1, 2, 3, 4, 5,])
serie.sum()

15

In [17]:
#Conteo de todos los elementos de la serie
serie.count()

5

In [18]:
#Obteniendo el valor máximo de la serie
serie.max()

5

In [19]:
#Obteniendo el valor mínimo de la serie
serie.min()

1

##### Funciones vectorizadas y Agregaciones con DataFrames

Tanto las funciones vectorizadas como las agregaciones pueden ser aplicadas a DataFrames completos.

Veamos primero las agregaciones. Al aplicar una agregación a un DataFrame, lo que obtenemos de regreso es el resultado de aplicar la función a cada una de las columnas (que al final de cuentas son Series, ¿lo recuerdas?). Por ejemplo, tenemos el siguiente DataFrame:

In [20]:
datos_productos = {
    "precio": [24, 64, 223, 38, 56, 12, 84],
    "cantidad_en_stock": [3, 6, 10, 2, 6, 45, 2],
    "productos_vendidos": [3, 45, 22, 76, 24, 8, 2]
}

indice = ["Pokemaster", "Cegatron", "Pikame Mucho", "Lazarillo de Tormes", "Stevie Wonder", "Needle", "El AyMeDuele"]

In [22]:
df = pd.DataFrame(datos_productos, index = indice)
df

precio  cantidad_en_stock  productos_vendidos
Pokemaster               24                  3                   3
Cegatron                 64                  6                  45
Pikame Mucho            223                 10                  22
Lazarillo de Tormes      38                  2                  76
Stevie Wonder            56                  6                  24
Needle                   12                 45                   8
El AyMeDuele             84                  2                   2

Mira qué pasa cuando le aplicamos la agregación sum:

In [24]:
df.sum()

precio                501
cantidad_en_stock      74
productos_vendidos    180
dtype: int64

pandas toma cada una de las columnas, suma todos los valores dentro de cada columna y nos regresa el resultado de las sumas en una nueva Serie, donde el índice son los nombres de las columnas en el DataFrame y los valores son las sumas.

También funciona con las demás agregaciones:

In [25]:
df.min()

precio                12
cantidad_en_stock      2
productos_vendidos     2
dtype: int64

In [26]:
df.max()

precio                223
cantidad_en_stock      45
productos_vendidos     76
dtype: int64

Ahora veamos qué pasa cuando aplicamos funciones vectorizadas a nuestro DataFrame:

In [27]:
df * 100

precio  cantidad_en_stock  productos_vendidos
Pokemaster             2400                300                 300
Cegatron               6400                600                4500
Pikame Mucho          22300               1000                2200
Lazarillo de Tormes    3800                200                7600
Stevie Wonder          5600                600                2400
Needle                 1200               4500                 800
El AyMeDuele           8400                200                 200

La función (x * 100) fue aplicada a cada uno de los elementos del DataFrame y obtuvimos un nuevo DataFrame con los resultados, ¿ves?

Cualquier función vectorizada que le apliquemos al DataFrame va a tener el mismo efecto:

In [28]:
df / 100

precio  cantidad_en_stock  productos_vendidos
Pokemaster             0.24               0.03                0.03
Cegatron               0.64               0.06                0.45
Pikame Mucho           2.23               0.10                0.22
Lazarillo de Tormes    0.38               0.02                0.76
Stevie Wonder          0.56               0.06                0.24
Needle                 0.12               0.45                0.08
El AyMeDuele           0.84               0.02                0.02

In [29]:
np.power(df, 2)

precio  cantidad_en_stock  productos_vendidos
Pokemaster              576                  9                   9
Cegatron               4096                 36                2025
Pikame Mucho          49729                100                 484
Lazarillo de Tormes    1444                  4                5776
Stevie Wonder          3136                 36                 576
Needle                  144               2025                  64
El AyMeDuele           7056                  4                   4

Ahora sí, estamos listos para aprender a lidiar con valores nulos (NaNs).

##### NaN's

Como ya dijimos, los NaN's (Not a Number) son valores nulos en nuestro conjunto de datos. Son valores que por alguna razón no se encuentran en nuestro dataset. A la hora de coleccionar nuestros datos, al momento de transcribirlos o de almacenarlos, algo pasó que algunos de esos datos faltan en el dataset final. Pandas está diseñado para lidiar con estos datos fácilmente.

Veamos un DataFrame con valores NaN (estamos usando el objeto de numpy np.nan para crear valores Nan):

In [46]:
datos = {
    "precio": [34, 54, np.nan, np.nan, 56, 12, 34],
    "cantidad_en_stock": [3, 6, 14, np.nan, 5, 2, 10],
    "productos_vendidos": [3, 45, 23, np.nan, 24, 6, np.nan]
}

indice = ["Pokemaster", "Cegatron", "Pikame Mucho", "Lazarillo de Tormes", "Stevie Wonder", "Needle", "El AyMeDuele"]

In [64]:
df_2 = pd.DataFrame(datos, index = indice)
df_2

precio  cantidad_en_stock  productos_vendidos
Pokemaster             34.0                3.0                 3.0
Cegatron               54.0                6.0                45.0
Pikame Mucho            NaN               14.0                23.0
Lazarillo de Tormes     NaN                NaN                 NaN
Stevie Wonder          56.0                5.0                24.0
Needle                 12.0                2.0                 6.0
El AyMeDuele           34.0               10.0                 NaN

Como puedes ver, a este DataFrame le faltan datos. Los valores NaN que contiene son valores nulos, y nos nos dicen nada útil, en realidad. Podríamos asumir cosas acerca de por qué faltan estos datos, pero no serían más que suposiciones.

Ahora, ¿qué pasa con estos datos nulos? Su presencia puede ser problemática para algunos de los análisis que queremos realizar.

En este caso nuestro dataset es muy pequeño y podemos visualizarlo todo de un solo vistazo. Pero vamos a imaginar que nuestro dataset es mucho más grande y que necesitamos saber si hay NaNs y cuántos. Para lograr esto podemos usar una función vectorizada llamada isna que checa cada valor en nuestro DataFrame, lo transforma en True cuando el valor es igual a NaN y a False cuando no lo es:

In [48]:
df_2.isna()

precio  cantidad_en_stock  productos_vendidos
Pokemaster            False              False               False
Cegatron              False              False               False
Pikame Mucho           True              False               False
Lazarillo de Tormes    True               True                True
Stevie Wonder         False              False               False
Needle                False              False               False
El AyMeDuele          False              False                True

Ok, ¿y ahora qué hacemos con esto? Podemos usar la agregación sum para hacer un conteo de nuestros valores nulos. Si recuerdas, sum sumaba todos los valores de cada columna y regresaba el total por columna. Si sumas valores booleanos, los Trues cuentan como 1 y los False cuentan como 0. Esto significa que aplicando la función, obtendremos el total de valores nulos en cada columna:

In [49]:
df_2.isna().sum()

precio                2
cantidad_en_stock     1
productos_vendidos    2
dtype: int64

De esta manera podemos saber si hay columnas que tienen demasiados valores nulos como para ser utilizadas.

También podemos obtener el número de NaNs que hay en cada fila pasándole un argumento a sum. Podemos indicarle a sum el eje en el cual queremos realizar la operación. En el caso de pandas eje se refiere a la dimensión de la estructura de datos. A estos ejes se les llama axis. Vamos a entender mejor los ejes más adelante cuando veamos aritmética de Series; por el momento basta con saber que si le pasamos axis=1 a sum nos regresa la suma de NaNs por índice:

In [50]:
df_2.isna().sum(axis = 1)

Pokemaster             0
Cegatron               0
Pikame Mucho           1
Lazarillo de Tormes    3
Stevie Wonder          0
Needle                 0
El AyMeDuele           1
dtype: int64

La decisión de qué hacer con los NaNs depende mucho del contexto. Vamos a ver 3 cosas básicas que podemos realizar para limpiar estos datos indeseables:

- Eliminar filas con NaNs
- Eliminar columnas con NaNs
- Llenar los NaNs con algún valor.

##### Eliminar filas con NaNs

Usando el método dropna, la eliminación de filas con NaNs se vuelve muy sencillo. Sólo basta con llamar dropna y todas las filas que contienen NaNs son eliminadas:


In [51]:
#No se modifica el dataframe original, si se requiere que persista, se debe reasignar
df_2.dropna()

precio  cantidad_en_stock  productos_vendidos
Pokemaster       34.0                3.0                 3.0
Cegatron         54.0                6.0                45.0
Stevie Wonder    56.0                5.0                24.0
Needle           12.0                2.0                 6.0

In [52]:
df_2_rows_dropped = df_2.dropna()
df_2_rows_dropped

precio  cantidad_en_stock  productos_vendidos
Pokemaster       34.0                3.0                 3.0
Cegatron         54.0                6.0                45.0
Stevie Wonder    56.0                5.0                24.0
Needle           12.0                2.0                 6.0

In [53]:
df_2

precio  cantidad_en_stock  productos_vendidos
Pokemaster             34.0                3.0                 3.0
Cegatron               54.0                6.0                45.0
Pikame Mucho            NaN               14.0                23.0
Lazarillo de Tormes     NaN                NaN                 NaN
Stevie Wonder          56.0                5.0                24.0
Needle                 12.0                2.0                 6.0
El AyMeDuele           34.0               10.0                 NaN

En caso de que queramos eliminar solamente las filas donde TODOS los valores sean NaN, podemos pasarle el argumento how='all':

In [54]:
df_2_rows_all_dropped = df_2.dropna(how = 'all')
df_2_rows_all_dropped

precio  cantidad_en_stock  productos_vendidos
Pokemaster       34.0                3.0                 3.0
Cegatron         54.0                6.0                45.0
Pikame Mucho      NaN               14.0                23.0
Stevie Wonder    56.0                5.0                24.0
Needle           12.0                2.0                 6.0
El AyMeDuele     34.0               10.0                 NaN

El valor default es how='any', que elimina las filas donde haya mínimo un NaN.

##### Eliminar columnas con NaNs

Ahora, ¿qué pasa si quisiéramos eliminar NaNs por columna? Vamos a agregar una columna a nuestro DataFrame que contenga puros NaNs:

In [65]:
df_2['descuento'] = np.nan
df_2

precio  cantidad_en_stock  productos_vendidos  descuento
Pokemaster             34.0                3.0                 3.0        NaN
Cegatron               54.0                6.0                45.0        NaN
Pikame Mucho            NaN               14.0                23.0        NaN
Lazarillo de Tormes     NaN                NaN                 NaN        NaN
Stevie Wonder          56.0                5.0                24.0        NaN
Needle                 12.0                2.0                 6.0        NaN
El AyMeDuele           34.0               10.0                 NaN        NaN

Si te fijas, sólo bastó con llamar df['descuento'] = np.nan para conseguir una columna completa de NaNs. Esto tiene que ver con la aritmética de Series. Cuando asigno un solo valor a una Serie, automáticamente toda la Serie toma ese valor.

Para eliminar columnas donde haya NaNs, llamamos también dropna pero con el argumento axis=1:

In [56]:
df_2.dropna(axis = 1)

Empty DataFrame
Columns: []
Index: [Pokemaster, Cegatron, Pikame Mucho, Lazarillo de Tormes, Stevie Wonder, Needle, El AyMeDuele]

Como aquí también el valor default es how=any, pandas elimina todas las columnas donde haya mínimo un NaN, que en este caso son todas. Para que elimine sólo las columnas donde todos los valores sean NaNs, hay que pasarle el argumento how = all:

In [57]:
df_2.dropna(axis = 1, how = 'all')

precio  cantidad_en_stock  productos_vendidos
Pokemaster             34.0                3.0                 3.0
Cegatron               54.0                6.0                45.0
Pikame Mucho            NaN               14.0                23.0
Lazarillo de Tormes     NaN                NaN                 NaN
Stevie Wonder          56.0                5.0                24.0
Needle                 12.0                2.0                 6.0
El AyMeDuele           34.0               10.0                 NaN

##### Llenar valores nulos con un valor

Digamos que por ahora nuestro DataFrame que se ve así:

In [66]:
df_2

precio  cantidad_en_stock  productos_vendidos  descuento
Pokemaster             34.0                3.0                 3.0        NaN
Cegatron               54.0                6.0                45.0        NaN
Pikame Mucho            NaN               14.0                23.0        NaN
Lazarillo de Tormes     NaN                NaN                 NaN        NaN
Stevie Wonder          56.0                5.0                24.0        NaN
Needle                 12.0                2.0                 6.0        NaN
El AyMeDuele           34.0               10.0                 NaN        NaN

Nuestra primera acción debería de ser eliminar las filas y columnas donde todos los valores sean NaN, porque no nos sirven de nada:

In [67]:
df_2_dropped = df_2.dropna(axis = 0, how = 'all')
df_2_dropped = df_2_dropped.dropna(axis = 1, how = 'all')

df_2_dropped

precio  cantidad_en_stock  productos_vendidos
Pokemaster       34.0                3.0                 3.0
Cegatron         54.0                6.0                45.0
Pikame Mucho      NaN               14.0                23.0
Stevie Wonder    56.0                5.0                24.0
Needle           12.0                2.0                 6.0
El AyMeDuele     34.0               10.0                 NaN

Ahora, ¿qué debemos hacer con valores nulos que nos quedan? Digamos que nuestro análisis más importante tiene que ver con la columna 'precio', entonces esa columna es muy importante que esté limpia. Pero digamos que nuestra 'productos_vendidos' no es tan importante. Tal vez si hay un NaN en productos vendidos podemos asumir que no hay ningún producto vendido hasta ahora. En ese caso, podríamos llenar el/los NaN de la columna 'productos_vendidos' con 0s. Eso se hace con el método fillna:

In [68]:
df_2_dropped['productos_vendidos'].fillna(0)

Pokemaster        3.0
Cegatron         45.0
Pikame Mucho     23.0
Stevie Wonder    24.0
Needle            6.0
El AyMeDuele      0.0
Name: productos_vendidos, dtype: float64

Seleccionamos la columna donde queremos llenar los NaNs con 0 y llamamos el métodos fillna(0). En este caso sólo estamos obteniendo de regreso la columna rellenada. Para tenerla en nuestro DataFrame podemos reasignarla a la misma columna:

In [69]:
df_2_dropped['productos_vendidos'] = df_2_dropped['productos_vendidos'].fillna(0)
df_2_dropped

precio  cantidad_en_stock  productos_vendidos
Pokemaster       34.0                3.0                 3.0
Cegatron         54.0                6.0                45.0
Pikame Mucho      NaN               14.0                23.0
Stevie Wonder    56.0                5.0                24.0
Needle           12.0                2.0                 6.0
El AyMeDuele     34.0               10.0                 0.0

Listo!

Ahora sí, podemos eliminar las filas que aún contengan NaNs porque sabemos que los NaNs que quedan son demasiados indeseables:

In [70]:
df_2_cleaned = df_2_dropped.dropna()
df_2_cleaned

precio  cantidad_en_stock  productos_vendidos
Pokemaster       34.0                3.0                 3.0
Cegatron         54.0                6.0                45.0
Stevie Wonder    56.0                5.0                24.0
Needle           12.0                2.0                 6.0
El AyMeDuele     34.0               10.0                 0.0

Y ya tenemos un dataset libre de valores nulos!

##### Aplicación en nuestro dataset original

Vamos a ver cómo funciona esto en nuestro dataset que teníamos al principio:

In [76]:
df.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/09/2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  4/02/2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.8014    144.9958  Northern Metropolitan   
1        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
2     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
3        NaN        Yarra  -37.8114    145.0116  Northern Metropolitan   
4     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

Primero hacemos conteo de NaNs:

In [77]:
df.isna().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             4344
Method               0
SellerG              0
Date                 0
Distance             8
Postcode             8
Bedroom2          4413
Bathroom          4413
Car               4413
Landsize          4796
BuildingArea     11123
YearBuilt        10389
CouncilArea       4444
Lattitude         4292
Longtitude        4292
Regionname           8
Propertycount        8
dtype: int64

Tenemos un total de filas de 19740, así que el hecho de que tengamos alrededor de 11000 valores NaNs en las columnas BuildingArea y YearBuilt no son una buena señal. Tal vez después cambiemos nuestra decisión, pero por el momento vamos a simplemente eliminarlas:

In [79]:
df_dropped = df.drop(columns = ['BuildingArea', 'YearBuilt'])
df_dropped.isna().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Price            4344
Method              0
SellerG             0
Date                0
Distance            8
Postcode            8
Bedroom2         4413
Bathroom         4413
Car              4413
Landsize         4796
CouncilArea      4444
Lattitude        4292
Longtitude       4292
Regionname          8
Propertycount       8
dtype: int64

Ok, ahora tenemos que decidir qué vamos a hacer con el resto de los NaNs. Presiento que no es muy grave no tener valores en 'Regionname', ya que es poco probable que usemos esa columna para nuestro análisis. Por lo tanto voy a llenar los NaNs con el valor Unknown:

In [80]:
df_dropped['Regionname'] = df_dropped['Regionname'].fillna('Unknow')
df_dropped.isna().sum()

Suburb              0
Address             0
Rooms               0
Type                0
Price            4344
Method              0
SellerG             0
Date                0
Distance            8
Postcode            8
Bedroom2         4413
Bathroom         4413
Car              4413
Landsize         4796
CouncilArea      4444
Lattitude        4292
Longtitude       4292
Regionname          0
Propertycount       8
dtype: int64

El resto de las columnas voy a considerarlas esenciales, así que vamos a eliminar todas las filas donde todavía tengamos NaNs:

In [81]:
df_dropped = df_dropped.dropna()
df_dropped.isna().sum()

Suburb           0
Address          0
Rooms            0
Type             0
Price            0
Method           0
SellerG          0
Date             0
Distance         0
Postcode         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
CouncilArea      0
Lattitude        0
Longtitude       0
Regionname       0
Propertycount    0
dtype: int64

Y ahora vemos cuántas filas nos han quedado:

In [82]:
df_dropped.shape

(11646, 19)

Si estas filas que nos quedan son suficientes o no, eso sólo lo sabremos continuando con el proceso.

##### Reindexando

Ahora, algo pasó con nuestro dataset después de eliminar los NaNs y es que nuestro índice ya no corresponde con el número de filas en nuestro dataset:

In [83]:
df_dropped

Suburb               Address  Rooms Type      Price Method  \
1        Abbotsford          85 Turner St      2    h  1480000.0      S   
2        Abbotsford       25 Bloomburg St      2    h  1035000.0      S   
4        Abbotsford          5 Charles St      3    h  1465000.0     SP   
5        Abbotsford      40 Federation La      3    h   850000.0     PI   
6        Abbotsford           55a Park St      4    h  1600000.0     VB   
...             ...                   ...    ...  ...        ...    ...   
19731    Whittlesea         30 Sherwin St      3    h   601000.0      S   
19734  Williamstown           87 Pasco St      3    h  1285000.0      S   
19737    Yarraville           2 Adeney St      2    h   750000.0     SP   
19738    Yarraville       54 Pentland Pde      6    h  2450000.0     VB   
19739    Yarraville  10/127 Somerville Rd      3    t   645000.0     SP   

             SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
1             Biggin   3/12/2016       2.5    3067.0       2.0       1.0  1.0   
2             Biggin   4/02/2016       2.5    3067.0       2.0       1.0  0.0   
4             Biggin   4/03/2017       2.5    3067.0       3.0       2.0  0.0   
5             Biggin   4/03/2017       2.5    3067.0       3.0       2.0  1.0   
6             Nelson   4/06/2016       2.5    3067.0       3.0       1.0  2.0   
...              ...         ...       ...       ...       ...       ...  ...   
19731            Ray  29/07/2017      35.5    3757.0       3.0       2.0  2.0   
19734            Jas  29/07/2017       6.8    3016.0       2.0       1.0  1.0   
19737  hockingstuart  29/07/2017       6.3    3013.0       3.0       2.0  2.0   
19738        Village  29/07/2017       6.3    3013.0       3.0       2.0  1.0   
19739            Jas  29/07/2017       6.3    3013.0       2.0       1.0  1.0   

       Landsize CouncilArea  Lattitude  Longtitude             Regionname  \
1         202.0       Yarra  -37.79960   144.99840  Northern Metropolitan   
2         156.0       Yarra  -37.80790   144.99340  Northern Metropolitan   
4         134.0       Yarra  -37.80930   144.99440  Northern Metropolitan   
5          94.0       Yarra  -37.79690   144.99690  Northern Metropolitan   
6         120.0       Yarra  -37.80720   144.99410  Northern Metropolitan   
...         ...         ...        ...         ...                    ...   
19731    1970.0  Manningham  -37.76311   145.10494      Northern Victoria   
19734    2010.0  Whittlesea  -37.68199   145.01744   Western Metropolitan   
19737    1999.0     Darebin  -37.75948   144.99615   Western Metropolitan   
19738    2011.0        Hume  -37.70322   144.88236   Western Metropolitan   
19739    1980.0        Hume  -37.69815   144.88019   Western Metropolitan   

       Propertycount  
1             4019.0  
2             4019.0  
4             4019.0  
5             4019.0  
6             4019.0  
...              ...  
19731         2170.0  
19734         6380.0  
19737         6543.0  
19738         6543.0  
19739         6543.0  

[11646 rows x 19 columns]

Esto sucede porque eliminamos filas pero las filas que se mantuvieron siguen teniendo el mismo índice que antes. Hay veces que eso es lo que queremos (cuando nuestro índice son, nombres, etiquetas, letras, etc), pero en este caso, nos convendría que nuestro índice coincidiera con la posición de la fila en el dataset. Podemos corregir esto usando el método reset_index de la siguiente manera:

In [84]:
df_dropped.reset_index().tail()

index        Suburb               Address  Rooms Type      Price  \
11641  19731    Whittlesea         30 Sherwin St      3    h   601000.0   
11642  19734  Williamstown           87 Pasco St      3    h  1285000.0   
11643  19737    Yarraville           2 Adeney St      2    h   750000.0   
11644  19738    Yarraville       54 Pentland Pde      6    h  2450000.0   
11645  19739    Yarraville  10/127 Somerville Rd      3    t   645000.0   

      Method        SellerG        Date  Distance  Postcode  Bedroom2  \
11641      S            Ray  29/07/2017      35.5    3757.0       3.0   
11642      S            Jas  29/07/2017       6.8    3016.0       2.0   
11643     SP  hockingstuart  29/07/2017       6.3    3013.0       3.0   
11644     VB        Village  29/07/2017       6.3    3013.0       3.0   
11645     SP            Jas  29/07/2017       6.3    3013.0       2.0   

       Bathroom  Car  Landsize CouncilArea  Lattitude  Longtitude  \
11641       2.0  2.0    1970.0  Manningham  -37.76311   145.10494   
11642       1.0  1.0    2010.0  Whittlesea  -37.68199   145.01744   
11643       2.0  2.0    1999.0     Darebin  -37.75948   144.99615   
11644       2.0  1.0    2011.0        Hume  -37.70322   144.88236   
11645       1.0  1.0    1980.0        Hume  -37.69815   144.88019   

                 Regionname  Propertycount  
11641     Northern Victoria         2170.0  
11642  Western Metropolitan         6380.0  
11643  Western Metropolitan         6543.0  
11644  Western Metropolitan         6543.0  
11645  Western Metropolitan         6543.0

In [85]:
df_dropped.shape

(11646, 19)

Como ves, el índice ahora corresponde con el número de filas que tenemos. El único problema es que ahora tenemos una columna llamada índice que contiene los índices anteriores. Una vez más, hay veces que queremos eso (cuando la información contenida ahí era relevante y no queremos deshacernos de ella), pero en este caso, en realidad no nos interesa mantener esa columna. Para resetear el índice y eliminarlo al mismo tiempo, usamos reset_index(drop=True):

In [91]:
df_reindexed = df_dropped.reset_index(drop = True)

In [93]:
df_reindexed.shape

(11646, 19)

In [92]:
df_reindexed.tail()

Suburb               Address  Rooms Type      Price Method  \
11641    Whittlesea         30 Sherwin St      3    h   601000.0      S   
11642  Williamstown           87 Pasco St      3    h  1285000.0      S   
11643    Yarraville           2 Adeney St      2    h   750000.0     SP   
11644    Yarraville       54 Pentland Pde      6    h  2450000.0     VB   
11645    Yarraville  10/127 Somerville Rd      3    t   645000.0     SP   

             SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
11641            Ray  29/07/2017      35.5    3757.0       3.0       2.0  2.0   
11642            Jas  29/07/2017       6.8    3016.0       2.0       1.0  1.0   
11643  hockingstuart  29/07/2017       6.3    3013.0       3.0       2.0  2.0   
11644        Village  29/07/2017       6.3    3013.0       3.0       2.0  1.0   
11645            Jas  29/07/2017       6.3    3013.0       2.0       1.0  1.0   

       Landsize CouncilArea  Lattitude  Longtitude            Regionname  \
11641    1970.0  Manningham  -37.76311   145.10494     Northern Victoria   
11642    2010.0  Whittlesea  -37.68199   145.01744  Western Metropolitan   
11643    1999.0     Darebin  -37.75948   144.99615  Western Metropolitan   
11644    2011.0        Hume  -37.70322   144.88236  Western Metropolitan   
11645    1980.0        Hume  -37.69815   144.88019  Western Metropolitan   

       Propertycount  
11641         2170.0  
11642         6380.0  
11643         6543.0  
11644         6543.0  
11645         6543.0

Listo!

##### Renombrando columnas

Además del índice, otros identificadores que tenemos que nos interesa mantener siempre limpios y claros son los nombres de nuestras columnas. En el caso de nuestro dataset, los nombres no son suficientemente homogéneos. Tenemos cosas como Regionname (segunda palabra con minúscula) y otras como CouncilArea (segunda palabra con mayúscula). Además, hay errores ortográficos (lattitude, longtitude). Y también que prefiero que los nombres sigan la convención de nombramiento de Python (snake_case). Voy a renombrar mis columnas para que sigan las mismas convenciones.

Normalmente renombramos columnas cuando los nombres:

- No son lo suficientemente claros
- No representan la información que contiene esa columna.
- Tienen información basura.
- Tienen errores ortográficos.
- No siguen la convención que hemos decidido que deberían de tener.
- Son demasiado largos o difíciles de escribir.

Cambiemos entonces nuestros nombres de columnas. Primero creamos un mapa de los nombres viejos a los nombres nuevos:


In [94]:
df_reindexed.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'CouncilArea', 'Lattitude', 'Longtitude', 'Regionname',
       'Propertycount'],
      dtype='object')

In [89]:
column_name_mapping = {
    'Suburb': 'suburb',
    'Address': 'address',
    'Rooms': 'rooms',
    'Type': 'type',
    'Price': 'price',
    'Method': 'method',
    'SellerG': 'seller_g',
    'Date': 'date',
    'Distance': 'distance',
    'Postcode': 'post_code',
    'Bedroom2': 'bedrooms',
    'Bathroom': 'bathroom',
    'Car': 'car',
    'Landsize': 'land_size',
    'CouncilArea': 'council_area',
    'Lattitude': 'latitude',
    'Longtitude': 'longitude',
    'Regionname': 'region_name',
    'Propertycount': 'property_count'
}

Y ahora usamos el método rename para cambiar los nombres:

In [95]:
df_renamed = df_reindexed.rename(columns = column_name_mapping)
df_renamed

suburb               address  rooms type      price method  \
0        Abbotsford          85 Turner St      2    h  1480000.0      S   
1        Abbotsford       25 Bloomburg St      2    h  1035000.0      S   
2        Abbotsford          5 Charles St      3    h  1465000.0     SP   
3        Abbotsford      40 Federation La      3    h   850000.0     PI   
4        Abbotsford           55a Park St      4    h  1600000.0     VB   
...             ...                   ...    ...  ...        ...    ...   
11641    Whittlesea         30 Sherwin St      3    h   601000.0      S   
11642  Williamstown           87 Pasco St      3    h  1285000.0      S   
11643    Yarraville           2 Adeney St      2    h   750000.0     SP   
11644    Yarraville       54 Pentland Pde      6    h  2450000.0     VB   
11645    Yarraville  10/127 Somerville Rd      3    t   645000.0     SP   

            seller_g        date  distance  post_code  bedrooms  bathroom  \
0             Biggin   3/12/2016       2.5     3067.0       2.0       1.0   
1             Biggin   4/02/2016       2.5     3067.0       2.0       1.0   
2             Biggin   4/03/2017       2.5     3067.0       3.0       2.0   
3             Biggin   4/03/2017       2.5     3067.0       3.0       2.0   
4             Nelson   4/06/2016       2.5     3067.0       3.0       1.0   
...              ...         ...       ...        ...       ...       ...   
11641            Ray  29/07/2017      35.5     3757.0       3.0       2.0   
11642            Jas  29/07/2017       6.8     3016.0       2.0       1.0   
11643  hockingstuart  29/07/2017       6.3     3013.0       3.0       2.0   
11644        Village  29/07/2017       6.3     3013.0       3.0       2.0   
11645            Jas  29/07/2017       6.3     3013.0       2.0       1.0   

       car  land_size council_area  latitude  longitude  \
0      1.0      202.0        Yarra -37.79960  144.99840   
1      0.0      156.0        Yarra -37.80790  144.99340   
2      0.0      134.0        Yarra -37.80930  144.99440   
3      1.0       94.0        Yarra -37.79690  144.99690   
4      2.0      120.0        Yarra -37.80720  144.99410   
...    ...        ...          ...       ...        ...   
11641  2.0     1970.0   Manningham -37.76311  145.10494   
11642  1.0     2010.0   Whittlesea -37.68199  145.01744   
11643  2.0     1999.0      Darebin -37.75948  144.99615   
11644  1.0     2011.0         Hume -37.70322  144.88236   
11645  1.0     1980.0         Hume -37.69815  144.88019   

                 region_name  property_count  
0      Northern Metropolitan          4019.0  
1      Northern Metropolitan          4019.0  
2      Northern Metropolitan          4019.0  
3      Northern Metropolitan          4019.0  
4      Northern Metropolitan          4019.0  
...                      ...             ...  
11641      Northern Victoria          2170.0  
11642   Western Metropolitan          6380.0  
11643   Western Metropolitan          6543.0  
11644   Western Metropolitan          6543.0  
11645   Western Metropolitan          6543.0  

[11646 rows x 19 columns]